src/react/Reconciler.js

```diff
@@ -0,0 +1,330 @@
+        value.args[i] = this._resolveDeeply(value.args[i], context, BranchStatus.NEW_BRANCH);
+      }
+      return value;
+    }
+    if (value instanceof ArrayValue) {
+      this._resolveFragment(value, context, branchStatus);
+      return value;
+    }
+    if (value instanceof ObjectValue && isReactElement(value)) {
+      let typeValue = Get(this.realm, value, "type");
+      let propsValue = Get(this.realm, value, "props");
+      let refValue = Get(this.realm, value, "ref");
+      if (typeValue instanceof StringValue) {
+        // terminal host component. Start evaluating its children.
+        if (propsValue instanceof ObjectValue) {
+          let childrenProperty = propsValue.properties.get("children");
+          if (childrenProperty) {
+            let childrenPropertyDescriptor = childrenProperty.descriptor;
+            // if the descriptor is undefined, the property is likely deleted, if it exists
+            // proceed to resolve the children
+            if (childrenPropertyDescriptor !== undefined) {
+              let childrenPropertyValue = childrenPropertyDescriptor.value;
+              invariant(childrenPropertyValue instanceof Value, `Bad "children" prop passed in JSXElement`);
+              let resolvedChildren = this._resolveDeeply(childrenPropertyValue, context, branchStatus);
+              childrenPropertyDescriptor.value = resolvedChildren;
+            }
+          }
+        }
+        return value;
+      }
+      // we do not support "ref" on <Component /> ReactElements
+      if (!(refValue instanceof NullValue)) {
+        this._assignBailOutMessage(value, `Bail-out: refs are not supported on <Components />`);
+        return value;
+      }
+      if (!(propsValue instanceof ObjectValue || propsValue instanceof AbstractValue)) {
+        return value;
+      }
+      if (!(typeValue instanceof ECMAScriptSourceFunctionValue)) {
+        return value;
+      }
+      try {
+        let { result, commitDidMountPhase } = this._renderAsDeepAsPossible(
+          typeValue,
+          propsValue,
+          context,
+          this._updateBranchStatus(branchStatus) // reduce branch status a value if possible
+        );
+        if (result === null || result instanceof UndefinedValue) {
+          return branchStatus === BranchStatus.NEW_BRANCH ? this._applyBranchedLogic(value) : value;
+        }
+        this.statistics.inlinedComponents++;
+        if (commitDidMountPhase !== null) {
+          commitDidMountPhase();
+        }
+        if (branchStatus === BranchStatus.NEW_BRANCH && result instanceof ObjectValue && isReactElement(result)) {
+          return this._applyBranchedLogic(result);
+        }
+        return result;
+      } catch (error) {
+        // assign a bail out message
+        if (error instanceof ExpectedBailOut) {
+          this._assignBailOutMessage(value, "Bail-out: " + error.message);
+        } else {
+          this._assignBailOutMessage(value, "Evaluation bail-out");
```

Comment:
See comment above. Let's not just sweep all kinds of internal errors under the carpet here...

Reply:
We need to resume the reconciliation process here though. Re-throwing here will abort the process entirely.

Reply:
@sebmarkbage this was based on your original logic, do you see another route? If there's an internal error, should we bail out of the tree entirely – given that it's very likely at the moment?

Reply:
Maybe we could also append the original error + message to the bail-out message.

Reply:
What if we do:

```js
} else if (error instanceof FatalError) {
  this._assignBailOutMessage(value, "Evaluation bail-out");
} else {
  throw error;
}
```

Since most errors that leads to bailout should probably be `FatalError` according to the new Prepack convention if I understand it correctly. E.g. due to abstract values.

If it errors for any other reason, then it is likely a bug?

It might error for other reasons like "completion" objects being thrown. E.g. if user code throws. We can add them to the whitelist of "approved" classes of errors that we expect as we find them.

That way arbitrary bugs in Prepack would kill the compiler instead of bailing out the optimization. So that we can fix them.